<html>
<head>
<!-- #BeginEditable "doctitle" -->
<title>Transcript</title>
<!-- #EndEditable -->
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
</head>
<script language="JavaScript">
<!--
function MM_openBrWindow(theURL,winName,features) { //v2.0
  window.open(theURL,winName,features);
}
//-->
</script>
<body bgcolor="#FFFFFF">
<table width="450" height="39" border="0">
  <tr>
    <td width="93" >
      <div align="left">
        <input type=submit value="Print" name="print" onClick="window.print(); return false;">
      </div>
    </td>
    <td width="384">
      <div align="right"><img src="nursa_logo_small_h30.gif" width="118" height="30"></div>
    </td>
    <td width="10">&nbsp;</td>
  </tr>
</table>
<table width="450" border="0">
  <tr>
    <td height="5"></td>
  </tr>
  <tr bgcolor="442359">
    <td height="3" bgcolor="006699"></td>
  </tr>
  <tr>
    <td height="7"></td>
  </tr>
  <tr> 
    <td> <p><b><font face="Verdana, Arial, Helvetica, sans-serif" size="2">Cloning 
        and Characterization of Coactivators</font></b></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">To 
        facilitate the cloning of coregulator cDNAs, a molecular screen was designed 
        based upon the yeast two-hybrid protein-protein interaction assay.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">First, a hybrid bait 
        bsm was created in which a receptor LBD was linked to the activation
        domain of a yeast transcription factor, Gal4.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Next, a library containing 
        millions of potential binding partners was created by fusing a human cDNA 
        library with the Gal4 activation domain.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">The 
        yeast cells contain a lacZ gene controlled by Gal4 DNA-binding sites. 
        Cells in which the lac Z gene is expressed will turn blue in the presence 
        of a chromogenic substrate, whereas cells in which it is not expressed 
        will remain white in appearance.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">In 
        cells in which the bait does not interact specifically with a library 
        protein, expression of the lacZ gene generally does not occur, since the 
        Gal4 transcription factor is not reassembled at the reporter gene promoter.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">In 
        cells in which the bait interacts specifically with a library protein, 
        the Gal4 factor is effectively reassembled, and touches off expression 
        of the lacZ gene. These cells appear blue on the plate.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Individual 
        clones are retrieved, their library plasmids recovered, and the cDNAs 
        sequenced to determine their identity. </font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Proteins 
        were initially designated as coactivators based on their ability to enhance 
        the function of receptors in HRE transient transfection assays, and to 
        relieve squelching between two competing receptors in these assays. </font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Screens 
        carried out in this way were the first to clone cDNAs encoding nuclear 
        receptor coregulator molecules, the TRIPs in David Moore&#8217;s laboratory, 
        and RIP140 also known as NRIP1, in Malcolm Parker&#8217;s laboratory.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">The 
        first authentic, common transcriptional coactivator was steroid receptor 
        coactivator 1, or SRC-1, first cloned in Bert O&#8217;Malley&#8217;s laboratory. 
        SRC-1 and two related proteins, GRIP-1, cloned first by Michael Stallcup, 
        and p/CIP, initially identified in Geoff Rosenfeld&#8217;s lab, make up 
        the SRC/p160 family of coactivators.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">The 
        SRC/p160 family is defined by the presence in the N-terminus of tandem 
        PAS and beta-H-L-H motifs;</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">A 
        centrally-located domain which binds the coactivators CBP and p300;</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">And 
        a C-terminal region which mediates interaction with the CARM-1 coactivator.<br>
        </font><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif"><br>
        As will be seen later, interactions between coactivators, and the control 
        of these interactions, are thought to be an important component in the 
        diversity of nuclear receptor action.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Malcolm 
        Parker&#8217;s laboratory was the first to show that a recurring structural 
        feature of many coactivators is an alpha-helical LXXLL motif, or nuclear 
        receptor box, present from a single to several copies in many coactivators, 
        which is implicated in their ligand-dependent recruitment by the receptor 
        AF-2. The SRC/p160 coactivator family, for example, has a conserved cluster 
        of NR boxes located in the central region of each member of the family</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Coactivators 
        can be categorized based upon their varied functional properties. Classes 
        of coactivators include:</font></p>
      <ul>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif"> 
          Acetyltransferases, such as members of the SRC/p160 family</font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif"> 
          Ubiquitin ligases, such as E6-AP</font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif"> 
          ATP-coupled chromatin remodeling complexes, such as the SWI/SNF/BRG-1 
          complex </font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif"> 
          Protein methylases, such as CARM-1 and PRMT-1</font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">RNA 
          transcripts, such as SRA</font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Cell 
          cycle regulators such as cdc 25B </font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">RNA 
          helicases such as p72 </font></li>
        <li><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">And 
          members of the TRAP/DRIP complex, which foster direct contact with components 
          of the basal transcription machinery</font></li>
      </ul>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">Click 
        on the book icons to access the respective PubMed links.</font></p>
      <p><font color="#000000" size="2" face="Verdana, Arial, Helvetica, sans-serif">To 
        read more about this topic, click on the &quot;Further Reading&quot; icon.</font></p>
      <p>&nbsp;</p></td>
  </tr>
</table>
<p>&nbsp;</p>
<p><font face="Arial, Helvetica, sans-serif" size="2"> </font></p>
</body>
</html>
